Sprite Sheet Editor with our own format castle-sprite-sheet

and3md opened the pull request:
Hi, This PR implements the basic sprite sheet editor with our new sprite sheet format. I have implemented all the basic operations + the ability to import atlases and import Starling files. However, this is the first version and has some shortcomings that can be checked in the [Sprite Sheet Editor project](https://github.com/castle-engine/castle-engine/projects/2). 

Linux:
![spritesheeteditor_linux](https://user-images.githubusercontent.com/18555708/108065433-48781c80-705e-11eb-86d4-54006546e2fe.png)
Windows:
![editorwindows](https://user-images.githubusercontent.com/18555708/108065444-4f069400-705e-11eb-8b98-d547d7567097.png)

## Castle Sprite Sheet Format

Bases on Starling, it currently differs only in two things:
- we can set FPS for each animation (so `anim1` can have 5.0, `anim2` can have 15.0 FPS)
- animation naming is always strict

## How to open editor:
Just click right on project files and choose `New sprite sheet...`: 

![sprite_sheet_editor_run](https://user-images.githubusercontent.com/18555708/108061825-27f99380-7059-11eb-9388-d8703be18934.png)

## List of implemented features
#### Animations:
- add
- remove (Ctrl + Del)
- rename (F2)
- move up (Ctrl + Up)
- move down (Ctrl + Down)
- move top
- move bottom
- preview

#### Frames:
- multiselect actions
  - add frames
  - remove frames (Ctrl + Del)
  - move left (Ctrl+Left)
  - move right (Ctrl+Right)
  - create new animation from selected frames (Ctrl+Shift+A)
  - import atlas image to animation
  - drag frames to move to other animation
  - drag with ctrl to copy to other animation
- single frame actions
  - move frame to top
  - move frame to end  
  - preview

#### Export options
  - set atlas size

#### Other:
  - import Starling file

## How to create sprite sheet
There are many ways to create sprite sheet:
1. Start from scratch - simply add animations, and frames
2. Import Starling file - when you import starling file you see import panel to choose options:
![importstarling](https://user-images.githubusercontent.com/18555708/108064123-86744100-705c-11eb-8277-72e98d167513.png)
After importing simply save your file as Castle Sprite Sheet.
3. If you only have sprites atlas, you can import it and next create animations by selecting proper frames and run `Create new animation from selected frames`.
![import_atlas](https://user-images.githubusercontent.com/18555708/108064695-55484080-705d-11eb-86ea-825ff2bc4f84.png)

![frames_actions](https://user-images.githubusercontent.com/18555708/108064714-5bd6b800-705d-11eb-888a-80e4c50dca42.png)

## Export options
Currently, a square PNG RGBA atlas is always created. But we will add/expand more options it in the future.

michaliskambi reviewed (changes requested):
Code quality review:

1. I did some small improvements and pushed them to your branch where it was easy :) See at my commits above, esp. the one where I improve XML saving: https://github.com/castle-engine/castle-engine/pull/248/commits/38813b26f37661baee2df5c01427b4fdd39d7961
 
2. I also resolved the conflict with latest master.

3. There are 2 small issues in the code, listed below.

4. The remaining bigger code-quality issue is something we already discussed: Now we have 2 loaders for Starling XML format, one in `castleinternalspritesheet.pas`, the other one in `x3dloadinternalstarling.pas`.

    - It seems that `x3dloadinternalstarling.pas` should be just removed, as `LoadCastleSpriteSheet` function just does the job.

    - Reason: we want to support Starling XML loading in the new sprite sheet editor, in the foreseeable future. For example to load Kenney assets in our sprite sheet editor. So whatever we do in `castleinternalspritesheet.pas` for our `.castle-sprite-sheet` will have to support loading "vanilla" `.starling-xml` too, anyway. 

    - So the separation `castleinternalspritesheet.pas` vs `x3dloadinternalstarling.pas` does not serve the original purpose (which was to allow simplifying `castleinternalspritesheet.pas`, by ignoring some possible Starling extensions and making room for only CGE-specific extensions).

michaliskambi commented:
Testing as a normal user, I found these issues:

5. TSpeedButton on the toolbar does not show a `Hint`, and thus it is not obvious what the 4 most prominent buttons on the toolbar do. I would say to make them just TBitBtn -- they'll be larger and uglier, but also more obvious this way.

6. Add a "Close" button
    - in the "File" menu
    - in the top-right corner of the window

7. I created new animation, added there 3 images with size 4000x3000:

    1. First problem: they are shown as "1 - 4000x3000, 3 - 4000x3000, 4 - 4000x3000". It seems that number 2 was skipped.

    Screenshot: 
![Zrzut ekranu z 2021-03-12 09-44-25](https://user-images.githubusercontent.com/1076978/110927877-7a5f6480-8326-11eb-9af7-be5eddd5784e.png)

    2. When I increase atlas size to 9000 to make them work, animation runs in the preview window, but sprite sheet scale was not adjusted, so I see a very zoomed-in version. Toggling the "preview" radio from "Animation" to "Frame" and then back to "Animation" sets proper sprite sheet scale.

    Screenshot with bug:
![Zrzut ekranu z 2021-03-12 09-44-42](https://user-images.githubusercontent.com/1076978/110927933-9236e880-8326-11eb-9a22-d1d5b15dfc70.png)


    Screenshot OK (after toggling preview to "Frame" and back to "Animation"):
![Zrzut ekranu z 2021-03-12 09-44-48](https://user-images.githubusercontent.com/1076978/110927947-96630600-8326-11eb-8a8a-a5a7fd505572.png)

8. Closing and opening the sprite sheet editor again, the previous value of "Maximum atlas size" is visible in the UI (like 9000, that I input above) but internally it seems the code considers the value 1024 as current, because if I add again the 3 images with size 4000x3000, I get a warning again "Maximum atlas size too small to fit all the frames. Increase it to at least 9000". Which seems wrong, as the UI already shows 9000, see screenshot.

    I think UI should reset the "Maximum atlas size" to 1024 when opening the dialog box (or even creating new sprite sheet).

    
![Zrzut ekranu z 2021-03-12 09-45-55](https://user-images.githubusercontent.com/1076978/110928054-ba264c00-8326-11eb-9d61-ea13c7cf7c9a.png)


9. Using "Import atlas" on this atlas file: https://github.com/castle-engine/demo-models/blob/master/starling/wolf/wolf.png , with columns = 4 and rows = 8, results in animation where some frames are vertically shifted.

    
![Zrzut ekranu z 2021-03-12 10-07-46](https://user-images.githubusercontent.com/1076978/110928084-c3171d80-8326-11eb-883d-a4eea42356da.png)


10. Importing Starling slime animation from https://github.com/castle-engine/demo-models/tree/master/starling/slime shows in the list the sprite being cut, and the animation is not visible. I need to move camera within the preview window to see it.

![Zrzut ekranu z 2021-03-12 11-20-13](https://user-images.githubusercontent.com/1076978/110928138-d32efd00-8326-11eb-8c3c-cbc285c8ca96.png)
![Zrzut ekranu z 2021-03-12 11-20-24](https://user-images.githubusercontent.com/1076978/110928142-d3c79380-8326-11eb-8d54-26977aeaf2bc.png)

michaliskambi commented:
To help with testing AD 7, I attach 3 trivial images 4000x3000 I used.
[large-images.zip](https://github.com/castle-engine/castle-engine/files/6129339/large-images.zip)

michaliskambi commented on tools/castle-editor/code/castle_editor.lpr:
```diff
@@ -9,19 +9,24 @@
   Application.Initialize;
   { When this code changes (e.g. new auto-created form is added),
     remember to also update ../../build-tool/data/custom_editor_template/castle_editor.lpr }
+  Application.CreateForm(TIcons, Icons);
```
Oh, and the changes to editor LPI also have to be replicated to `../../build-tool/data/custom_editor_template/castle_editor.lpi`,

michaliskambi commented:
11. One more code quality note: I see you added build modes to the editor lpi file, `tools/castle-editor/code/castle_editor.lpi`. So we now have 3 modes (default, debug, release). I'd say to either

    1. Remove `debug` and `release`, for simplicity. The editor itself should not have any time-intensive code (that is all in CGE package), so the `default` build mode may be enough.
    2. Or, we remove `default` mode, and dedicate ourselves to always testing editor in both `debug` and `release` modes. The important thing here is that if we decide to have 2 modes -- then both modes must be tested.

and3md commented:
@michaliskambi Thank you for testing and notes! I found and fixed some nasty bugs :) I think all issues are fixed now, please retest. One exception is wolf.png (point 9) because it has not constant row height and this simple atlas importer works only when you have all rows and cols with the same size. I think we can add more modes and some intelligence to it in the future. I think adding that is out of scope of this PR.

>     1. One more code quality note: I see you added build modes to the editor lpi file, `tools/castle-editor/code/castle_editor.lpi`. So we now have 3 modes (default, debug, release). I'd say to either
>        
>        1. Remove `debug` and `release`, for simplicity. The editor itself should not have any time-intensive code (that is all in CGE package), so the `default` build mode may be enough.
>        2. Or, we remove `default` mode, and dedicate ourselves to always testing editor in both `debug` and `release` modes. The important thing here is that if we decide to have 2 modes -- then both modes must be tested.

OK, I removed Debug and Release. This looks like a more optimal solution.

michaliskambi commented:
Thank you! All sounds OK. I'll retest ASAP :)

and3md commented:
Added extra fix for animation preview when animations have different sizes :)

michaliskambi commented:
Thank you for this gigantic work!

Things left over:

- I found one additional bug when retesting: https://trello.com/c/ZqWpO9pP/146-importing-slime-starling-results-in-atlas-size-128-should-be-312

- Making a news announcement about it :) I'll try to make it this weekend.

- Finish https://github.com/castle-engine/cge-platformer-demo (also means applying 2 other pending PRs from @and3md related to it).
